File: sdks/typescript/src/constraints.ts
// Constraint evaluation — mirrors the Go reference constraints.go exactly.
// Every semantic must produce the same verdict for the same inputs, or
// cross-language conformance fails.

import type {
  Constraint,
  DelegationCert,
  VerifierContext,
} from "./types.js";

/**
 * Run every Constraint on cert against the caller-supplied VerifierContext.
 * Returns null iff all pass; an error string otherwise.
 * Fail-closed: an unknown Type or a constraint whose required context field
 * is absent causes rejection.
 */
export function evaluateConstraints(
  cert: DelegationCert,
  ctx: VerifierContext,
  nowSec: number,
): string | null {
  const list = cert.constraints ?? [];
  for (let i = 0; i < list.length; i++) {
    const err = evaluateConstraint(list[i]!, cert.cert_id, ctx, nowSec);
    if (err !== null) {
      return `constraint[${i}] (${list[i]!.type}): ${err}`;
    }
  }
  return null;
}

function evaluateConstraint(
  c: Constraint,
  certID: string,
  ctx: VerifierContext,
  nowSec: number,
): string | null {
  switch (c.type) {
    case "geo_circle": {
      if (ctx.current_lat === undefined || ctx.current_lon === undefined) {
        return "constraint_unverifiable: no current location in context";
      }
      if (c.lat === undefined || c.lon === undefined || c.radius_m === undefined) {
        return "malformed geo_circle: lat/lon/radius_m required";
      }
      const d = haversineMeters(ctx.current_lat, ctx.current_lon, c.lat, c.lon);
      if (d > c.radius_m) {
        return `outside allowed radius: ${d.toFixed(1)}m > ${c.radius_m.toFixed(1)}m`;
      }
      return null;
    }

    case "geo_polygon": {
      if (ctx.current_lat === undefined || ctx.current_lon === undefined) {
        return "constraint_unverifiable: no current location in context";
      }
      if (!c.points || c.points.length < 3) {
        return "polygon has fewer than 3 points";
      }
      // v1 polygon is defined over equirectangular projection — correct
      // for small regions, incorrect for anti-meridian-crossing shapes.
      // Fail closed rather than silently return wrong answers. SPEC §5.7.2
      // documents the v1 small-region limitation; geodesic semantics are v2.
      if (polygonSpansAntimeridian(c.points)) {
        return "geo_polygon spans >180° longitude — v1 semantics are undefined for anti-meridian-crossing polygons (see SPEC §5.7.2)";
      }
      if (!pointInPolygon(ctx.current_lat, ctx.current_lon, c.points)) {
        return "outside allowed polygon";
      }
      return null;
    }

    case "geo_bbox": {
      if (ctx.current_lat === undefined || ctx.current_lon === undefined) {
        return "constraint_unverifiable: no current location in context";
      }
      if (
        c.min_lat === undefined ||
        c.min_lon === undefined ||
        c.max_lat === undefined ||
        c.max_lon === undefined
      ) {
        return "malformed geo_bbox: min/max lat/lon required";
      }
      // Format mirrors Go's %.6f — cross-SDK error_reason parity requires it.
      if (ctx.current_lat < c.min_lat || ctx.current_lat > c.max_lat) {
        return `latitude ${ctx.current_lat.toFixed(6)} outside [${c.min_lat.toFixed(6)}, ${c.max_lat.toFixed(6)}]`;
      }
      // Anti-meridian-aware longitude check (SPEC §5.7.2).
      //   min_lon <= max_lon → ordinary bbox, lon must be in [min_lon, max_lon]
      //   min_lon >  max_lon → bbox wraps the 180° meridian (e.g.,
      //     min_lon=170, max_lon=-170 = "from 170°E through 180 to -170°W")
      //     — inside iff lon >= min_lon OR lon <= max_lon.
      if (c.min_lon <= c.max_lon) {
        if (ctx.current_lon < c.min_lon || ctx.current_lon > c.max_lon) {
          return `longitude ${ctx.current_lon.toFixed(6)} outside [${c.min_lon.toFixed(6)}, ${c.max_lon.toFixed(6)}]`;
        }
      } else {
        if (ctx.current_lon < c.min_lon && ctx.current_lon > c.max_lon) {
          return `longitude ${ctx.current_lon.toFixed(6)} outside wrapped [${c.min_lon.toFixed(6)}, ${c.max_lon.toFixed(6)}]`;
        }
      }
      const hasAlt = (c.min_alt_m ?? 0) !== 0 || (c.max_alt_m ?? 0) !== 0;
      if (hasAlt) {
        const alt = ctx.current_alt_m ?? Number.NaN;
        if (Number.isNaN(alt)) {
          return "constraint_unverifiable: no altitude in context but bbox has altitude bounds";
        }
        if (alt < (c.min_alt_m ?? 0) || alt > (c.max_alt_m ?? 0)) {
          return `altitude ${alt.toFixed(1)}m outside [${(c.min_alt_m ?? 0).toFixed(1)}m, ${(c.max_alt_m ?? 0).toFixed(1)}m]`;
        }
      }
      return null;
    }

    case "time_window": {
      if (!c.tz || !c.start || !c.end) {
        return "malformed time_window: tz/start/end required";
      }
      const start = parseHHMM(c.start);
      const end = parseHHMM(c.end);
      if (start === null) return `bad start time: ${c.start}`;
      if (end === null) return `bad end time: ${c.end}`;
      // Get current local time in the declared zone.
      let parts: Intl.DateTimeFormatPart[];
      try {
        parts = new Intl.DateTimeFormat("en-US", {
          timeZone: c.tz,
          hour: "2-digit",
          minute: "2-digit",
          hour12: false,
        }).formatToParts(new Date(nowSec * 1000));
      } catch {
        return `unknown timezone "${c.tz}"`;
      }
      let h = Number(parts.find((p) => p.type === "hour")?.value ?? -1);
      const m = Number(parts.find((p) => p.type === "minute")?.value ?? -1);
      if (h < 0 || m < 0) return "time_window: could not resolve local time";
      // Some ICU builds format midnight as 24:00 for en-US + hour12:false.
      // Go reports 00:00, and conformance requires stable error text.
      if (h === 24) h = 0;
      const cur = h * 60 + m;
      if (start <= end) {
        if (cur < start || cur > end) {
          return `current ${fmtHHMM(h, m)} outside [${c.start}, ${c.end}] ${c.tz}`;
        }
      } else {
        // Wrapping window (e.g. 22:00 to 06:00).
        if (cur < start && cur > end) {
          return `current ${fmtHHMM(h, m)} outside wrapped [${c.start}, ${c.end}] ${c.tz}`;
        }
      }
      return null;
    }

    case "max_speed_mps": {
      if (ctx.current_speed_mps === undefined) {
        return "constraint_unverifiable: no current speed in context";
      }
      if (c.max_mps === undefined) {
        return "malformed max_speed_mps: max_mps required";
      }
      if (ctx.current_speed_mps > c.max_mps) {
        return `speed ${ctx.current_speed_mps.toFixed(2)}mps exceeds max ${c.max_mps.toFixed(2)}mps`;
      }
      return null;
    }

    case "max_amount": {
      if (
        ctx.requested_amount === undefined ||
        ctx.requested_currency === undefined
      ) {
        return "constraint_unverifiable: no requested amount in context";
      }
      if (c.max_amount === undefined || !c.currency) {
        return "malformed max_amount: max_amount/currency required";
      }
      if (ctx.requested_currency !== c.currency) {
        return `currency mismatch: requested "${ctx.requested_currency}", constraint "${c.currency}"`;
      }
      if (ctx.requested_amount > c.max_amount) {
        return `amount ${ctx.requested_amount.toFixed(2)} ${ctx.requested_currency} exceeds max ${c.max_amount.toFixed(2)} ${c.currency}`;
      }
      return null;
    }

    case "max_rate": {
      if (!ctx.invocations_in_window) {
        return "constraint_unverifiable: no rate counter in context";
      }
      if (!c.count || !c.window_s || c.count <= 0 || c.window_s <= 0) {
        return "malformed max_rate: count and window_s must be positive";
      }
      const got = ctx.invocations_in_window(certID, c.window_s);
      if (got >= c.count) {
        return `rate limit exceeded: ${got} invocations in last ${c.window_s}s (max ${c.count})`;
      }
      return null;
    }

    default:
      // Sentinel prefix so the verifier can route to
      // identity_status=constraint_unknown. Matches the Go reference's
      // unknownConstraintError shape.
      return `constraint_unknown: unknown constraint type "${String((c as { type: unknown }).type)}"`;
  }
}

// ---- helpers ----

function haversineMeters(
  lat1: number,
  lon1: number,
  lat2: number,
  lon2: number,
): number {
  const earthRadiusM = 6371000;
  const rad = Math.PI / 180;
  const dLat = (lat2 - lat1) * rad;
  const dLon = (lon2 - lon1) * rad;
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(lat1 * rad) * Math.cos(lat2 * rad) * Math.sin(dLon / 2) ** 2;
  return 2 * earthRadiusM * Math.asin(Math.min(1, Math.sqrt(a)));
}

/**
 * Returns true if the polygon's longitudes span more than 180°. The only
 * way to span more than half the globe in longitude is to cross the
 * anti-meridian, and equirectangular ray-casting can't handle that — so
 * we refuse these polygons up-front (fail-closed) rather than silently
 * compute wrong inclusion. Mirrors Go's polygonSpansAntimeridian.
 */
function polygonSpansAntimeridian(points: [number, number][]): boolean {
  if (points.length < 3) return false;
  let minLon = points[0]![1];
  let maxLon = points[0]![1];
  for (const p of points) {
    if (p[1] < minLon) minLon = p[1];
    if (p[1] > maxLon) maxLon = p[1];
  }
  return maxLon - minLon > 180;
}

function pointInPolygon(
  lat: number,
  lon: number,
  poly: [number, number][],
): boolean {
  let inside = false;
  const n = poly.length;
  for (let i = 0, j = n - 1; i < n; j = i++) {
    const [yi, xi] = poly[i]!; // [lat, lon]
    const [yj, xj] = poly[j]!;
    if (
      ((yi > lat) !== (yj > lat)) &&
      lon < ((xj - xi) * (lat - yi)) / (yj - yi) + xi
    ) {
      inside = !inside;
    }
  }
  return inside;
}

function parseHHMM(s: string): number | null {
  if (s.length !== 5 || s[2] !== ":") return null;
  const h = Number(s.slice(0, 2));
  const m = Number(s.slice(3, 5));
  if (!Number.isFinite(h) || !Number.isFinite(m)) return null;
  if (h < 0 || h > 23 || m < 0 || m > 59) return null;
  return h * 60 + m;
}

function fmtHHMM(h: number, m: number): string {
  return `${h.toString().padStart(2, "0")}:${m.toString().padStart(2, "0")}`;
}

File: sdks/rust/src/constraints.rs
//! Constraint evaluation — mirrors Go's constraints.go exactly.
//!
//! Every semantic must produce the same verdict for the same inputs, or
//! cross-language conformance fails.

use crate::types::{Constraint, ConstraintEvaluator, DelegationCert, VerifierContext};

use chrono::{DateTime, Utc};
use chrono_tz::Tz;
use std::collections::HashMap;

/// Run every Constraint on cert against the caller-supplied VerifierContext.
/// Return `Ok(())` iff all pass; an error string otherwise.
/// Fail-closed: unknown Type or missing required context field causes rejection.
/// (SPEC §17.7) Unknown built-in types fall through to `ext_evaluators`
/// before failing closed.
pub fn evaluate_constraints<'a>(
    cert: &DelegationCert,
    ctx: &VerifierContext,
    now_sec: i64,
    ext_evaluators: Option<&HashMap<String, Box<dyn ConstraintEvaluator + 'a>>>,
) -> Result<(), String> {
    for (i, c) in cert.constraints.iter().enumerate() {
        let mut err = evaluate_constraint(c, &cert.cert_id, ctx, now_sec);
        if let Err(ref msg) = err {
            if msg.starts_with("constraint_unknown:") {
                if let Some(map) = ext_evaluators {
                    if let Some(ev) = map.get(&c.kind) {
                        err = ev.evaluate(c, &cert.cert_id, ctx, now_sec);
                    }
                }
            }
        }
        if let Err(e) = err {
            return Err(format!("constraint[{}] ({}): {}", i, c.kind, e));
        }
    }
    Ok(())
}

fn evaluate_constraint(
    c: &Constraint,
    cert_id: &str,
    ctx: &VerifierContext,
    now_sec: i64,
) -> Result<(), String> {
    match c.kind.as_str() {
        "geo_circle" => {
            let (lat, lon) = match (ctx.current_lat, ctx.current_lon) {
                (Some(a), Some(b)) => (a, b),
                _ => return Err("constraint_unverifiable: no current location in context".into()),
            };
            let d = haversine_meters(lat, lon, c.lat, c.lon);
            if d > c.radius_m {
                return Err(format!(
                    "outside allowed radius: {:.1}m > {:.1}m",
                    d, c.radius_m
                ));
            }
            Ok(())
        }
        "geo_polygon" => {
            let (lat, lon) = match (ctx.current_lat, ctx.current_lon) {
                (Some(a), Some(b)) => (a, b),
                _ => return Err("constraint_unverifiable: no current location in context".into()),
            };
            if c.points.len() < 3 {
                return Err("polygon has fewer than 3 points".into());
            }
            // v1 polygon is defined over equirectangular projection — correct
            // for small regions, incorrect for anti-meridian-crossing shapes.
            // Fail closed rather than silently return wrong answers. SPEC §5.7.2
            // documents the v1 small-region limitation; geodesic semantics are v2.
            if polygon_spans_antimeridian(&c.points) {
                return Err("geo_polygon spans >180° longitude — v1 semantics are undefined for anti-meridian-crossing polygons (see SPEC §5.7.2)".into());
            }
            if !point_in_polygon(lat, lon, &c.points) {
                return Err("outside allowed polygon".into());
            }
            Ok(())
        }
        "geo_bbox" => {
            let (lat, lon) = match (ctx.current_lat, ctx.current_lon) {
                (Some(a), Some(b)) => (a, b),
                _ => return Err("constraint_unverifiable: no current location in context".into()),
            };
            if lat < c.min_lat || lat > c.max_lat {
                // Format must match Go's %.6f (fixed 6 decimals) so cross-SDK
                // error_reason strings are byte-identical. Don't change.
                return Err(format!(
                    "latitude {:.6} outside [{:.6}, {:.6}]",
                    lat, c.min_lat, c.max_lat
                ));
            }
            // Anti-meridian-aware longitude check (SPEC §5.7.2).
            //   min_lon <= max_lon → ordinary bbox, lon must be in [min_lon, max_lon]
            //   min_lon >  max_lon → bbox wraps the 180° meridian (e.g.,
            //     min_lon=170, max_lon=-170 = "from 170°E through 180 to -170°W")
            //     — inside iff lon >= min_lon OR lon <= max_lon.
            if c.min_lon <= c.max_lon {
                if lon < c.min_lon || lon > c.max_lon {
                    return Err(format!(
                        "longitude {:.6} outside [{:.6}, {:.6}]",
                        lon, c.min_lon, c.max_lon
                    ));
                }
            } else if lon < c.min_lon && lon > c.max_lon {
                return Err(format!(
                    "longitude {:.6} outside wrapped [{:.6}, {:.6}]",
                    lon, c.min_lon, c.max_lon
                ));
            }
            let has_alt = c.min_alt_m != 0.0 || c.max_alt_m != 0.0;
            if has_alt {
                let alt = ctx.current_alt_m.ok_or_else(|| {
                    "constraint_unverifiable: no altitude in context but bbox has altitude bounds"
                        .to_string()
                })?;
                if alt < c.min_alt_m || alt > c.max_alt_m {
                    return Err(format!(
                        "altitude {:.1}m outside [{:.1}m, {:.1}m]",
                        alt, c.min_alt_m, c.max_alt_m
                    ));
                }
            }
            Ok(())
        }
        "time_window" => {
            if c.tz.is_empty() || c.start.is_empty() || c.end.is_empty() {
                return Err("malformed time_window: tz/start/end required".into());
            }
            let start =
                parse_hhmm(&c.start).ok_or_else(|| format!("bad start time: {}", c.start))?;
            let end = parse_hhmm(&c.end).ok_or_else(|| format!("bad end time: {}", c.end))?;
            let zone: Tz =
                c.tz.parse()
                    .map_err(|_| format!("unknown timezone \"{}\"", c.tz))?;
            let utc = DateTime::<Utc>::from_timestamp(now_sec, 0)
                .ok_or_else(|| "time_window: invalid now_sec".to_string())?;
            let local = utc.with_timezone(&zone);
            let cur = (local.format("%H").to_string().parse::<i32>().unwrap_or(0)) * 60
                + local.format("%M").to_string().parse::<i32>().unwrap_or(0);
            if start <= end {
                if cur < start || cur > end {
                    return Err(format!(
                        "current {} outside [{}, {}] {}",
                        local.format("%H:%M"),
                        c.start,
                        c.end,
                        c.tz
                    ));
                }
            } else {
                // Wrapping window (e.g. 22:00 to 06:00).
                if cur < start && cur > end {
                    return Err(format!(
                        "current {} outside wrapped [{}, {}] {}",
                        local.format("%H:%M"),
                        c.start,
                        c.end,
                        c.tz
                    ));
                }
            }
            Ok(())
        }
        "max_speed_mps" => {
            let speed = ctx.current_speed_mps.ok_or_else(|| {
                "constraint_unverifiable: no current speed in context".to_string()
            })?;
            if speed > c.max_mps {
                return Err(format!(
                    "speed {:.2}mps exceeds max {:.2}mps",
                    speed, c.max_mps
                ));
            }
            Ok(())
        }
        "max_amount" => {
            let amount = ctx.requested_amount.ok_or_else(|| {
                "constraint_unverifiable: no requested amount in context".to_string()
            })?;
            let req_ccy = ctx.requested_currency.as_deref().ok_or_else(|| {
                "constraint_unverifiable: no requested currency in context".to_string()
            })?;
            if req_ccy != c.currency {
                return Err(format!(
                    "currency mismatch: requested \"{}\", constraint \"{}\"",
                    req_ccy, c.currency
                ));
            }
            if amount > c.max_amount {
                return Err(format!(
                    "amount {:.2} {} exceeds max {:.2} {}",
                    amount, req_ccy, c.max_amount, c.currency
                ));
            }
            Ok(())
        }
        "max_rate" => {
            let counter = ctx
                .invocations_in_window
                .as_ref()
                .ok_or_else(|| "constraint_unverifiable: no rate counter in context".to_string())?;
            if c.count <= 0 || c.window_s <= 0 {
                return Err("malformed max_rate: count and window_s must be positive".into());
            }
            let got = counter(cert_id, c.window_s);
            if got >= c.count {
                return Err(format!(
                    "rate limit exceeded: {} invocations in last {}s (max {})",
                    got, c.window_s, c.count
                ));
            }
            Ok(())
        }
        other => Err(format!(
            // Sentinel prefix routes the verifier to identity_status=constraint_unknown.
            "constraint_unknown: unknown constraint type \"{}\"",
            other
        )),
    }
}

// ---- helpers ----

fn haversine_meters(lat1: f64, lon1: f64, lat2: f64, lon2: f64) -> f64 {
    let earth_radius_m = 6371000.0_f64;
    let rad = std::f64::consts::PI / 180.0;
    let d_lat = (lat2 - lat1) * rad;
    let d_lon = (lon2 - lon1) * rad;
    let a = (d_lat / 2.0).sin().powi(2)
        + (lat1 * rad).cos() * (lat2 * rad).cos() * (d_lon / 2.0).sin().powi(2);
    2.0 * earth_radius_m * a.sqrt().min(1.0).asin()
}

/// Returns true if the polygon's longitudes span more than 180°. The only
/// way to span more than half the globe in longitude is to cross the
/// anti-meridian, and equirectangular ray-casting can't handle that — so
/// we refuse these polygons up-front (fail-closed). Mirrors Go's
/// polygonSpansAntimeridian.
fn polygon_spans_antimeridian(points: &[[f64; 2]]) -> bool {
    if points.len() < 3 {
        return false;
    }
    let mut min_lon = points[0][1];
    let mut max_lon = points[0][1];
    for p in points {
        if p[1] < min_lon {
            min_lon = p[1];
        }
        if p[1] > max_lon {
            max_lon = p[1];
        }
    }
    (max_lon - min_lon) > 180.0
}

fn point_in_polygon(lat: f64, lon: f64, poly: &[[f64; 2]]) -> bool {
    let mut inside = false;
    let n = poly.len();
    let mut j = n - 1;
    for i in 0..n {
        let (yi, xi) = (poly[i][0], poly[i][1]); // lat, lon
        let (yj, xj) = (poly[j][0], poly[j][1]);
        if ((yi > lat) != (yj > lat)) && lon < (xj - xi) * (lat - yi) / (yj - yi) + xi {
            inside = !inside;
        }
        j = i;
    }
    inside
}

fn parse_hhmm(s: &str) -> Option<i32> {
    let b = s.as_bytes();
    if b.len() != 5 || b[2] != b':' {
        return None;
    }
    let h: i32 = s[0..2].parse().ok()?;
    let m: i32 = s[3..5].parse().ok()?;
    if !(0..=23).contains(&h) || !(0..=59).contains(&m) {
        return None;
    }
    Some(h * 60 + m)
}
